diff -p ./closedShop.pm ../closedShop.new/closedShop.pm
*** ./closedShop.pm	2004-07-13 16:56:42.000000000 -0400
--- ../closedShop.new/closedShop.pm	2004-10-04 11:59:21.000000000 -0400
*************** sub closedShop::Initialize
*** 78,85 ****
  	  $cat_template $item_template %Languages $allcat_template
          /;
  	
! 	$varsfile = "closedShop.vars";
! 	$gcfile = "closedShop_gc.vars";
  	$MaxCats = 20;
  	
  	LoadVariables("$varsfile",$_[1]);
--- 78,85 ----
  	  $cat_template $item_template %Languages $allcat_template
          /;
  	
! 	$varsfile = "vars/closedShop.vars";
! 	$gcfile = "vars/closedShop_gc.vars";
  	$MaxCats = 20;
  	
  	LoadVariables("$varsfile",$_[1]);
Only in .: cS.diff
diff -p ./Install.pl ../closedShop.new/Install.pl
*** ./Install.pl	2004-07-13 17:07:44.000000000 -0400
--- ../closedShop.new/Install.pl	2004-10-04 11:59:31.000000000 -0400
*************** use Fcntl ':flock';
*** 50,57 ****
  my $query = new CGI;
  my $program_name = "Install.pl";
  
! my $varsfile = "closedShop.vars";
! my $gcfile = "closedShop_gc.vars";
  my @Languages = ("English","Tongan");
  	   
  my $font = qq| <font face="arial, verdana" size="3"> |;
--- 50,57 ----
  my $query = new CGI;
  my $program_name = "Install.pl";
  
! my $varsfile = "vars/closedShop.vars";
! my $gcfile = "vars/closedShop_gc.vars";
  my @Languages = ("English","Tongan");
  	   
  my $font = qq| <font face="arial, verdana" size="3"> |;
*************** your cgi-bin 777) and go back and try ag
*** 209,222 ****
  sub InstallScreen
  {
      use closedShop;
!     open(FILE,">test_cs_file") || return closedShop::ErrorMessage("
          <h2>closedShop Installation Error</h2>
          You do not have permission to
  	create files in this folder. <BR>You need to fix this problem before
  	continuing. <BR>Try chmod'ing this folder 777 if need be or
          make sure that the web server user can write files here. <BR>If none of that
          makes sense, ask someone for help so that closedShop can write to this folder");
!     unlink("test_cs_file");
  
  	my $message = shift;
  	Header();
--- 209,222 ----
  sub InstallScreen
  {
      use closedShop;
!     open(FILE,">vars/test_cs_file") || return closedShop::ErrorMessage("
          <h2>closedShop Installation Error</h2>
          You do not have permission to
  	create files in this folder. <BR>You need to fix this problem before
  	continuing. <BR>Try chmod'ing this folder 777 if need be or
          make sure that the web server user can write files here. <BR>If none of that
          makes sense, ask someone for help so that closedShop can write to this folder");
!     unlink("vars/test_cs_file");
  
  	my $message = shift;
  	Header();
